extensions/amp-date-display/0.1/amp-date-display.js

```diff
@@ -0,0 +1,269 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpEvents} from '../../../src/amp-events';
+import {Services} from '../../../src/services';
+import {createCustomEvent} from '../../../src/event-helper';
+import {isExperimentOn} from '../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {removeChildren} from '../../../src/dom';
+import {user} from '../../../src/log';
+
+/** @const {string} */
+const DEFAULT_LOCALE = 'en';
+
+/** @const {number} */
+const DEFAULT_OFFSET_SECONDS = 0;
+
+export class AmpDateDisplay extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const {function(!Element)} */
+    this.boundRendered_ = this.rendered_.bind(this);
+
+    /** @private {string} */
+    this.datetime_ = '';
+
+    /** @private {number} */
+    this.timestampSeconds_ = 0;
+
+    /** @private {number} */
+    this.timestampMiliseconds_ = 0;
+
+    /** @private {string} */
+    this.displayIn_ = '';
+
+    /** @private {number} */
+    this.offsetSeconds_ = 0;
+
+    /** @private {string} */
+    this.locale_ = '';
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    if (!isExperimentOn(this.getWin(), 'amp-date-display')) {
+      user.warn('Experiment `amp-date-display` is not turned on.');
+      return;
+    }
+
+    // Store this in buildCallback() because `this.element` sometimes
+    // is missing attributes in the constructor.
+
+    // Note: One of datetime, timestamp-ms, timestamp-seconds is required.
+
+    /** @private {string} */
+    this.datetime_ = this.element.getAttribute('datetime') || '';
+
+    /** @private {number} */
+    this.timestampSeconds_ = Number(
+        this.element.getAttribute('timestamp-seconds')
+    );
+
+    /** @private {number} */
+    this.timestampMiliseconds_ = Number(
+        this.element.getAttribute('timestamp-ms')
+    );
+
+    /** @private {string} */
+    this.displayIn_ = this.element.getAttribute('display-in') || '';
+
+    /** @private {number} */
+    this.offsetSeconds_ =
+      Number(this.element.getAttribute('offset-seconds')) ||
+      DEFAULT_OFFSET_SECONDS;
+
+    /** @private {string} */
+    this.locale_ = this.element.getAttribute('locale') || DEFAULT_LOCALE;
+
+    /** @private {Object} */
+    const data = this.getDataForTemplate_();
+
+    this.templates_
+        .findAndRenderTemplate(this.element, data)
+        .then(this.boundRendered_);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @return {Object}
```

Comment:
I added the definitions. Knowing Flow and TypeScript, this took me some time 😅

BTW, is there a way to avoid the duplication? I tried `@extend` but it does not work on typedefs.